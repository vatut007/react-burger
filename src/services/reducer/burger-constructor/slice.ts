import { PayloadAction, createSlice, nanoid } from "@reduxjs/toolkit";
import {
  IngredientConstructorList,
  type Ingredient
} from "../../../types/ingredient";

export const burgerConstructorSlice = createSlice({
  name: "burgerConstructor",
  reducerPath: "burgerConstructor",
  initialState: {
    bun: null as Ingredient | null,
    ingredients: [] as IngredientConstructorList,
    ingredientCount: {} as Record<string,number>
  },
  reducers: {
    addIngredient: {
      reducer(
        state,
        action: PayloadAction<{ cartItemId: string; ingredient: Ingredient }>,
      ) {
        if (action.payload.ingredient.type === "bun") {
          state.bun = action.payload.ingredient;
        } else {
          state.ingredients.push({
            ...action.payload.ingredient,
            cart_item_id: action.payload.cartItemId,
          });
          if (action.payload.ingredient._id in state.ingredientCount){
            state.ingredientCount[action.payload.ingredient._id]++
          } else {
            state.ingredientCount[action.payload.ingredient._id]=1
          }
        }
      },
      prepare({ ingredient }: { ingredient: Ingredient }) {
        const cartItemId = nanoid();
        return { payload: { cartItemId, ingredient } };
      },
    },
    removeIngredient(state, action: PayloadAction<{ cartItemId: string, ingeredient_id: string }>) {
      state.ingredients = state.ingredients.filter(
        (item) => item.cart_item_id !== action.payload.cartItemId,
      );
      state.ingredientCount[action.payload.ingeredient_id]--
    },
    moveIngredient(
      state,
      action: PayloadAction<{ oldIndex: string; newIndex: number }>,
    ) {
      const index = state.ingredients.findIndex(
        (item) => item.cart_item_id === action.payload.oldIndex,
      );
      const [ingredient] = state.ingredients.splice(index, 1);
      state.ingredients.splice(action.payload.newIndex, 0, ingredient);
    },
  },
});
