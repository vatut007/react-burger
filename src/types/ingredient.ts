export type IngredientType = "main" | "sauce" | "bun";

export type Ingredient = {
  _id: string;
  name: string;
  type: IngredientType;
  proteins: number;
  fat: number;
  carbohydrates: number;
  calories: number;
  price: number;
  image: string;
  image_mobile: string;
  image_large: string;
  __v: number;
};

export type IngredientConstructor = {
  cart_item_id: string;
  _id: string;
  name: string;
  type: IngredientType;
  proteins: number;
  fat: number;
  carbohydrates: number;
  calories: number;
  price: number;
  image: string;
  image_mobile: string;
  image_large: string;
  __v: number;
};

export type IngredientList = Ingredient[];

export type IngredientConstructorList = IngredientConstructor[];

export type IngredientListResponse = { success: boolean; data: IngredientList };

export const types: IngredientType[] = ["bun", "sauce", "main"];
