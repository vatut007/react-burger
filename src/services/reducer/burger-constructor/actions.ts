import { burgerConstructorSlice } from "./slice";

export const { addIngredient, removeIngredient, moveIngredient } =
  burgerConstructorSlice.actions;
