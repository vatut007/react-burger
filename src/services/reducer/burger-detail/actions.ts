import { ingredientDetailSlice } from "./slice";

export const { selectIngredient } = ingredientDetailSlice.actions;
