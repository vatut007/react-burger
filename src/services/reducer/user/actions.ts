import { userSlice } from "./slice";

export const { addAccessToken, addUser, addRefreshToken, clearUser } =
  userSlice.actions;
