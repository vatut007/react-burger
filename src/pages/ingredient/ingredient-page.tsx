import { useParams } from "react-router-dom";
import { AppHeader } from "../../components/app-header/app-header";
import { IngredientDetail } from "../../components/ingredient-detail/ingredient-detail";
import { useDispatch } from "react-redux";
import { useEffect } from "react";
import { selectIngredient } from "../../services/reducer/burger-detail/actions";
import {
  ingredientSelectors,
  useGetAllIngredientQuery,
} from "../../services/api/api-slice";
import styles from "./ingredient-pages.module.css";

export function IngredientPage() {
  const { ingredientId } = useParams();
  const {
    data: ingredientEntities,
    error,
    isLoading,
  } = useGetAllIngredientQuery(undefined);
  const dispath = useDispatch();
  useEffect(() => {
    if (ingredientEntities && ingredientId) {
      const ingredient = ingredientSelectors.selectById(
        ingredientEntities,
        ingredientId,
      );
      if (ingredient) {
        dispath(selectIngredient({ ingredient }));
      }
    }
  }, [ingredientEntities]);
  return (
    <div className={styles.div}>
      <div className={styles.detailDiv}>
        <IngredientDetail />
      </div>
    </div>
  );
}
