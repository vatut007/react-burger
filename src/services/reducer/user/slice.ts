import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { User } from "../../../types/user";

export const userSlice = createSlice({
  name: "user",
  reducerPath: "user",
  initialState: {
    user: null as User | null,
    accessToken: null as string | null,
    refreshToken: null as string | null,
  },
  reducers: {
    addUser(state, action: PayloadAction<User>) {
      state.user = {
        email: action.payload.email,
        name: action.payload.name,
      };
    },
    addAccessToken(state, action: PayloadAction<{ accessToken: string }>) {
      state.accessToken = action.payload.accessToken;
    },
    addRefreshToken(state, action: PayloadAction<{ refreshToken: string }>) {
      state.refreshToken = action.payload.refreshToken;
    },
    clearUser(state) {
      (state.accessToken = null),
        (state.refreshToken = null),
        (state.user = null);
    },
  },
});
