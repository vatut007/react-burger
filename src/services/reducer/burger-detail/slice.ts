import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Ingredient } from "../../../types/ingredient";

export const ingredientDetailSlice = createSlice({
    name: 'ingedientDetail',
    reducerPath: 'ingedientDetail',
    initialState: {ingredient:null as Ingredient | null},
    reducers:{
        selectIngredient(state, action: PayloadAction<{ ingredient: Ingredient }>) {
            state.ingredient = action.payload.ingredient
        },
    }
})