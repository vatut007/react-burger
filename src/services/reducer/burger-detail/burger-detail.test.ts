import { afterEach, beforeEach, expect, test, vi } from "vitest";
import { ingredientDetailSlice } from "./slice";
import { selectIngredient } from "./actions";
import { Ingredient } from "../../../types/ingredient";

const { reducer, getInitialState } = ingredientDetailSlice;

const getTestIngredient = (): Ingredient => ({
  _id: "643d69a5c3f7b9001cfa093d",
  name: "Флюоресцентная булка R2-D3",
  type: "bun",
  proteins: 44,
  fat: 26,
  carbohydrates: 85,
  calories: 643,
  price: 988,
  image: "https://code.s3.yandex.net/react/code/bun-01.png",
  image_mobile: "https://code.s3.yandex.net/react/code/bun-01-mobile.png",
  image_large: "https://code.s3.yandex.net/react/code/bun-01-large.png",
  __v: 0,
});

test("select Ingredient", () => {
  expect(
    reducer(
      getInitialState(),
      selectIngredient({
        ingredient: getTestIngredient(),
      }),
    ),
  ).toMatchSnapshot();
});
