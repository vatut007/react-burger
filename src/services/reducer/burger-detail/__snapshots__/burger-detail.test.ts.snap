// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`select Ingredient 1`] = `
{
  "ingredient": {
    "__v": 0,
    "_id": "643d69a5c3f7b9001cfa093d",
    "calories": 643,
    "carbohydrates": 85,
    "fat": 26,
    "image": "https://code.s3.yandex.net/react/code/bun-01.png",
    "image_large": "https://code.s3.yandex.net/react/code/bun-01-large.png",
    "image_mobile": "https://code.s3.yandex.net/react/code/bun-01-mobile.png",
    "name": "Флюоресцентная булка R2-D3",
    "price": 988,
    "proteins": 44,
    "type": "bun",
  },
}
`;
